### PR TITLE
fix: change sampling to without replacement

### PR DESCRIPTION
**Pull request by thatlittleboy**
## Change 1: change sampling to without replacement

cf. slundberg#2738

---

`shap.utils.sample()` is primarily used in our `datasets` module and notebooks to generate a subset of the original data. E.g. [here](https://github.com/dsgibbons/shap/blob/e24b7019cda8889376b72353f6ad8bdcb1db0bcc/shap/datasets.py#L47).

However, the current implementation uses [`sklearn`'s resample](https://github.com/scikit-learn/scikit-learn/blob/72a604975102b2d93082385d7a5a7033886cc825/sklearn/utils/__init__.py#L476) which uses `replace=True` by default.

I think this is unexpected / unwanted behaviour. This PR changes to using sklearn `shuffle()` instead of `resample()` which is just the same as setting `replace=False`.

## Other minor changes

* add tests for `shap.utils.sample`
* use `len()` in cases where there is no `.shape` attribute, e.g. for ordinary python lists.

### CODE REVIEW

**Comment by dsgibbons** [comment]
LGTM. I agree, sampling without replacement makes more sense.

**Review by dsgibbons — APPROVED** [review]

**Comment by thatlittleboy** [comment]
Thanks for the confirmation. I left the PR as draft for the moment as I want to add some tests for this `sample` function, to prevent similar regression in the future.

Will request for review again later this week.

**Comment by thatlittleboy** [comment]
Alright, PR ready for review. Tests for `sample()` have passed in CI.

@dsgibbons

**Comment by dsgibbons** [comment]
LGTM!
